### ISSUE TITLE
Raise the default maxHttpHeaderSize on tomcat to avoid all Solr Search Queries failing

### ISSUE THREAD

**Issue by toonconvents** [Alfresco/alfresco-sdk#440]
## Expected Behavior
When configuring several facets with buckets, I expect the Alfresco search functionality to keep on working.

## Current Behavior
Queries to Solr are http request that have a very large header size because of all possible facet query filters that are sent. There is a point at which there are 'too many' facet buckets configured and the queries surpass the default Tomcat maxHttpHeaderSize of 8k. From then on all queries fail and always return 0 results in the Share search results

## Possible Solution
Configure the built-in tomcat with a higher maxHttpHeaderSize on the port 8080 connector. (i.e. 65536)
Also configure a bigger packetSize on the AJP (port 8009) connector. (i.e. 65536)

## My Environment
* Alfresco SDK version used: 2.2.0
* Alfresco version used:  Enterprise 5.1.2

**Comment by ohej** [comment]
There is no easy way for the SDK to support this. The only work around I see is that you would need to provide a custom `server.xml` file. The Tomcat plugin supports this via `<serverXml>`. 

It seems you can set the property `maven.tomcat.serverXml` to point to a custom one where you could specify this. Can you try it out? See http://tomcat.apache.org/maven-plugin-2.0/tomcat7-maven-plugin/run-mojo.html#serverXml for more info (all though there isn't a whole more than what I just said..)

**Comment by toonconvents** [comment]
Hi Ole,

I tried that but then the plugin completely ignores the webapp blocks in the config so that introduces another problem of not being able to rely on the maven dependencies for wars...
In the end I had to actually create a patched version of the tomcat maven plugin with a hardcoded set of the limits to 65536 and install that in our local maven repo for us to be able to use the search in development. Not ideal at all...

**Comment by ohej** [comment]
Hi Toon

Hmm. Have you tried this with SDK3? Or is this a 2.x thing? If you could create a sample project with the serverXml it would make it easier to try and debug, but unfortunately this might be one area where we will have a tough time if it's a limitation/bug in the tomcat plugin.

**Comment by toonconvents** [comment]
This is on 2.2.0. 
By now I think it is a problem with the tomcat-maven-plugin. It doesn't allow you to configure this without doing the whole server.xml override thing.

**Comment by ohej** [comment]
But wouldn't that be OK to do? - overriding the whole server.xml?

**Comment by toonconvents** [comment]
Sorry for the late answer. I was on holiday.
As I said, overriding the server.xml introduces the problem of not being able to rely on the maven dependencies for wars. So unless I'm missing something this would be far from ideal...

**Comment by toonconvents** [comment]
Hi @ohej ,
In the meantime, what we did as a workaround was create our own version of the tomcat maven plugin and push it to our maven repo under a special version number. Now that I'm upgrading our development project to SDK3 it seems I cant override the version of the tomcat-maven-plugin used anymore and this bug is still occurring...
Any idea as to a viable workaround? Will we need to replace the version used by the alfresco maven plugin in our maven repo or is there a simple way to tell the alfresco maven plugin which version of the tomcat-maven-plugin to use?
Thanks in advance.

**Comment by vojkog** [comment]
Hi, @toonconvents,
I have a similar problem, that forced me to use custom server.xml. 
As tomcat plugin docs state (for using _maven.tomcat.serverXml_): 
> "if you use this you must configure in this file your webapp paths"  

So I ended up with something like this in my custom server.xml:
```xml
<Context path="/alfresco" docBase="../../platform-war" />   
<Context path="/share" docBase="../../share-war" />  
<Context path="/myapp" docBase="external path to myapp.war" />  
```  
Not entirely happy with this solution, so I would love to hear other options.  
Thanks.

**Comment by mauro1855** [comment]
I'm using SDK 3.0.1 for Alfresco 5.2 and I have the same problem.
I set up maybe around 6 or 7 custom facets and the search doesn't work anymore. After a lot of research I realized that it was related to the maxHttpHeaderSize and I can't seem to figure out a way to increase this in embeded tomcat.

**Comment by mauro1855** [comment]
@vojkog I couldn't get your solution to work in SDK 3.0.1.
If I do that I start having exceptions in the logs:

`Cannot load JDBC driver class '${alfresco.db.datasource.class}'
java.lang.ClassNotFoundException: ${alfresco.db.datasource.class}
	at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1702)
	at org.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1547)
`

**Comment by marsv024** [comment]
We are also seeing this problem, anyone have an updated solution for this on SDK3?

**Comment by ohej** [comment]
After a lot of investigation, there is no easy way to fix this. 

We have been studying different alternatives to solve this issue inside the code of the Alfresco SDK:

1.  **Tomcat startup hook**. Try to link some code to the startup of Tomcat to modify the configuration of the HTTP Connector to modify the maxHttpHeaderSize value to a higher value. Neither the Tomcat Maven Plugin nor Tomcat offers a hook to execute code on the server startup. Apart from that, it seems that any change in the configuration of the connector would require a server restart. 

2. **Configuration update via JMX.** Tomcat offers the possibility to modify some parts of its configuration via JMX. Sadly, the properties we need to modify are not included in the offering of Tomcat JMX. 

3. **Custom LifecycleListener.** Tomcat allows the configuration of lifecycle listeners (https://tomcat.apache.org/tomcat-7.0-doc/config/listeners.html) over Server, Engine, Host or Context. These listeners receive events before the startup of the server. That would be a good moment to modify the configuration of the maxHttpHeaderSize. The main issue is that listener must be registered in the server.xml file of the Tomcat server in order to make it executable within the Tomcat server lifecycle. If we could modify the server.xml file we would simply update the value of the property instead of registering the listener. 

4. **Try to modify the default server.xml file of the Tomcat.** The Tomcat Maven Plugin starts up an embedded version of the Tomcat server. Due to that, the server.xml file doesn't exist physically. That avoids any strategy of modifying the content of that file. And anyway, the Tomcat Maven Plugin doesn't offer any hook to include actions during the startup of the embedded Tomcat server. 

5. **Configure an external server.xml file.** One alternative that provides the Tomcat Maven Plugin is configuring an external server.xml file using the property serverXml. The main issue with this approach is that in that case, as it is stated in the documentation of the plugin, "you must configure in this file your webapp paths". That means that, if the Alfresco Maven Plugin wants to offer an automatic and transparent solution to this issue, it would have to programmatically create the server.xml file based on the configuration of the different web applications to deploy. That would add more complexity to an already complex Alfresco Maven Plugin. 

6. **Enable a property to set the version of the Tomcat Maven Plugin to use.** This alternative would require the integrator to modify the Tomcat Maven Plugin to include the maxHttpHeaderSize (not hard to do), deploy a new version of the Tomcat plugin in their corporate Maven repository and use the new parameter to use the custom version of the Tomcat Maven Plugin. Viable
Create a patched version of the tomcat maven plugin. This approach is similar to the previous one. But in this case the "patch" of the Tomcat Maven Plugin is done by ourselves and the patched version of the plugin is deployed to the Alfresco Public Maven repo. This way, the Alfresco Maven Plugin can be modified to use the patched version of the plugin that fixes the value of the maxHttpHeaderSize property. 

Sadly, there is not an easy solution for this issue that can be implemented in the own code of the Alfresco Maven plugin. 

We've decided to go for option 6. We open the possibility to set the version of the Tomcat Maven Plugin to use, but it is the client company who would eventually patch the Tomcat Maven Plugin. This has been merged in #528.

Moving forward with SDK 4, since we don't rely on the Tomcat plugin anymore this won't be an issue.
